horace_core/sqw/@PixelData/PixelData.m

```diff
@@ -327,41 +440,110 @@
+        detector_index = obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :);
     end
 
     function obj = set.detector_idx(obj, detector_indices)
-       obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :) = detector_indices;
+        obj.data(obj.FIELD_INDEX_MAP_('detector_idx'), :) = detector_indices;
     end
 
     function detector_index = get.energy_idx(obj)
-       detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        detector_index = obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :);
     end
 
     function obj = set.energy_idx(obj, energies)
         obj.data(obj.FIELD_INDEX_MAP_('energy_idx'), :) = energies;
-     end
+    end
 
     function signal = get.signal(obj)
-       signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        signal = obj.data(obj.FIELD_INDEX_MAP_('signal'), :);
     end
 
     function obj = set.signal(obj, signal)
         obj.data(obj.FIELD_INDEX_MAP_('signal'), :) = signal;
-     end
+    end
 
     function variance = get.variance(obj)
-       variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
+        obj = obj.load_first_page_if_data_empty_();
+        variance = obj.data(obj.FIELD_INDEX_MAP_('variance'), :);
     end
 
     function obj = set.variance(obj, variance)
         obj.data(obj.FIELD_INDEX_MAP_('variance'), :) = variance;
-     end
+    end
 
     function num_pix = get.num_pixels(obj)
-        num_pix = size(obj.data, 2);
+        if isempty(obj.f_accessor_)
+            num_pix = size(obj.data, 2);
+        else
+            num_pix = obj.f_accessor_.npixels;
+        end
+    end
+
+    function file_path = get.file_path(obj)
+        file_path = obj.file_path_;
+    end
+
+    function page_size = get.page_size(obj)
+        % The number of pixels that are held in the current page. If no pixels
+        % are held, this returns the number of pixels that will be loaded into
+        % the next page.
+        %
+        % If "num_pixels in file" < "number of pixels that can fit in memory size",
+        % return num_pixels - this avoids ever reading past end of file
+        if isempty(obj.data)
+            num_bytes_in_val = 8;  % pixel data stored in memory as a double
+            num_bytes_in_pixel = num_bytes_in_val*obj.PIXEL_BLOCK_COLS_;
+            page_size = floor(obj.page_memory_size_/num_bytes_in_pixel);
+            if page_size > obj.num_pixels
+                page_size = obj.num_pixels;
+            end
+        else
+            page_size = size(obj.data, 2);
+        end
+    end
+
+end
+
+methods (Access = private)
+
+    function obj = init_from_file_accessor_(obj, f_accessor)
+        % Initialise a PixelData object from a file accessor
+        obj.f_accessor_ = f_accessor;
+        obj.file_path_ = fullfile(obj.f_accessor_.filepath, ...
+                                  obj.f_accessor_.filename);
+        pg_size = obj.page_size;
```

Comment:
unnecessary local variable `pg_size`

Reply:
Why only load the first page if it hold all the data?

Reply:
`obj.page_size` is a dependent variable, so having the local variable saved a function call.

If all the data can be held in one page we may as well load that in. If the data needs multiple pages, we hold off on loading until we know which page needs to be loaded. This will be most useful when we implement `get_data_at_pix_index`